Logic/Path/ball_path.py

```diff
@@ -6,210 +6,189 @@
+            # Line defines the path between two balls it is assumed to be blocked if
+            # the distance less than two ball radii
 
-                if angle < self.min_angle or angle > self.max_angle:
-                    return False
+            is_intercepted = self.vectors.line_intercept_circle(line, a_ball, int(lookup.BALL_DIAMETER))
 
-                if angle < 180:
-                    angle_weight = (180 - angle) * 1000
-                else:
-                    angle_weight = (angle - 180) * 1000
-
-            distance = self.distance_from_two_points(all_objects[start_index], all_objects[end_index])
-            graph.add_edge(start_index, end_index, angle_weight + distance)
-
-        return is_valid
-
-    def is_path_ocluded(self, start_position, end_position, path_indexes, all_objects):
-        '''Determine if path is ocluded by other balls'''
+            if i not in exclude_indices and is_intercepted:
+                return False
+        
+        return True
+    
+    @staticmethod
+    def is_possible_shot(white, target_ball, target_hole):
+        lower_target_ball = (target_ball[0] - lookup.BALL_DIAMETER, target_ball[1] - lookup.BALL_DIAMETER)
+        upper_target_ball = (target_ball[0] - lookup.BALL_DIAMETER, target_ball[1] - lookup.BALL_DIAMETER)
 
-        line = self.line_equation_from_two_points(start_position, end_position)
+        lower_target_hole = (target_hole[0] - lookup.BALL_DIAMETER, target_hole[1] - lookup.BALL_DIAMETER)
+        upper_target_hole = (target_hole[0] - lookup.BALL_DIAMETER, target_hole[1] - lookup.BALL_DIAMETER)
 
-        for ball_index, a_ball in enumerate(all_objects):
-            if (ball_index not in path_indexes and len(a_ball) == 3):
-                if not self.line_intercept_circle(line, a_ball, lookup.BALL_DIAMETER):
-                    return False
+        not_valid_x = lower_target_ball[0] < white[0] < upper_target_hole[0] or lower_target_hole[0] < white[0] < upper_target_ball[0]
+        not_valid_y = lower_target_ball[1] < white[1] < upper_target_hole[1] or lower_target_hole[1] < white[1] < upper_target_ball[1]
 
-        return True
+        return not (not_valid_x or not_valid_y)
 
     @staticmethod
     def get_balls_index(balls, ball_colour):
         '''Find balls which have a particular colour'''
 
-        ball_colour_indexes = []
+        ball_colour_indices = []
 
         for i, _ in enumerate(balls):
             if balls[i][2] is ball_colour:
-                ball_colour_indexes.append(i)
+                ball_colour_indices.append(i)
 
         if (ball_colour is BallColour.White or ball_colour is BallColour.Black):
-            if ball_colour_indexes:
-                return ball_colour_indexes[0]
+            if ball_colour_indices:
+                return ball_colour_indices[0]
 
-        return ball_colour_indexes
+        return ball_colour_indices
+    
+    def get_target_holes(self):
+        '''Responsible for finding the target holes'''
```

Comment:
Should 'target holes' be explained further?

Reply:
Done